tests/test_api_lib_policy.py

```diff
@@ -2048,3 +2049,65 @@ def test_17_pin_change(self):
         # finally delete policy
         delete_policy("pol1")
         delete_policy("pol2")
+
+    def test_18_challenge_text_header(self):
+        # This looks like a validate/check request, that triggers a challenge
+        builder = EnvironBuilder(method='POST',
+                                 data={'user': "hans",
+                                       "pass": "pin"},
+                                 headers={})
+
+        env = builder.get_environ()
+        # Set the remote address so that we can filter for it
+        env["REMOTE_ADDR"] = "10.0.0.1"
+        g.client_ip = env["REMOTE_ADDR"]
+        req = Request(env)
+
+        # Set a policy for the header
+        set_policy(name="pol_header",
+                   scope=SCOPE.AUTH,
+                   action="{0!s}=These are your options:<ul>".format(ACTION.CHALLENGETEXT_HEADER))
+        # Set a policy for the footer
+        set_policy(name="pol_footer",
+                   scope=SCOPE.AUTH,
+                   action="{0!s}=Happy authenticating!".format(ACTION.CHALLENGETEXT_FOOTER))
+        g.policy_object = PolicyClass()
+
+        req.all_data = {
+            "user": "hans",
+            "pass": "pin"}
+        req.User = User()
+
+        # We do an html list
+        res = {"jsonrpc": "2.0",
+               "result": {"status": True,
+                          "value": False},
+               "version": "privacyIDEA test",
+               "id": 1,
+               "detail": {"message": "enter the HOTP from your email, "
+                                     "enter the HOTP from your email, "
+                                     "enter the TOTP from your SMS",
+                          "messages": ["enter the HOTP from your email",
+                                       "enter the HOTP from your email",
+                                       "enter the TOTP from your SMS"],
+                          "multi_challenge": ["chal1", "chal2", "chal3"]}}
+        resp = jsonify(res)
+
+        r = mangle_challenge_response(req, resp)
+        message = r.json.get("detail", {}).get("message")
+        self.assertEqual(message, "These are your options:<ul><li>enter the TOTP from your SMS</li>\n<li>enter the HOTP from your email</li>\nHappy authenticating!")
```

Comment:
I think we cannot be sure of the order of elements here, because we perform ``list(set(messages))`` on the messages. But I guess we'll see if the test randomly fails :-)

Reply:
Ah actually, just noticed that the test has already failed on Travis :-) https://travis-ci.org/privacyidea/privacyidea/jobs/493770918#L2131
So I guess we'll need a different assertion here -- something like ``self.assertIn("<li>...</li>", message)``?

Reply:
This would not detect double entries. We could return `sorted(set(messages))` which returns a sorted list.

Reply:
This is interesting.
This way it could even be sorted by adding some hidden tags.
I like this "sorted" approach.

Will do.

Reply:
done in a94cc884